File: Java/MaximumSubarray.java

public class MaximumSubarray {

	public static void main(String[] args) {

	}
	
	public static int maxSubArray(int[] nums) {
        // time complexity: O(n)
		// space complexity: O(1)
		
		int maxSub = nums[0];
        int curSub = 0;
        
        for (int right = 0; right < nums.length; right++) {
        	if (curSub < 0)
        		curSub = nums[right];
        	else
        		curSub += nums[right];
        	
        	if (curSub > maxSub)
        		maxSub = curSub;
        }
        
        return maxSub;
    }

}

File: C++/maximumSubarray.cpp
#include <vector>

class Solution {
public:
    int maxSubArray(std::vector<int>& nums)
    {
        // time complexity: O(n)
        // space complexity: O(1)
        
        int max = INT32_MIN;
        int sub = 0;

        for (int i : nums)
        {
            if (sub < 0)
            {
                sub = 0;
            }
            sub += i;
            max = std::max(max, sub);
        }
        return max;
    }
};
